pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/GroupedStages.java

```diff
@@ -0,0 +1,177 @@
+ * stages.
+ *
+ * It is important to understand that this class assumes all stages that are stored belong to the same query plan
+ * and therefore their stage ids are unique. It also assumes that the same stage instances are being used when
+ * methods like {@link #containsStage(MailboxSendNode)} are called.
+ *
+ * The original reason to have this class was to group equivalent stages together, although it can be used for other
+ * purposes.
+ *
+ * Although the only implementation provided so far ({@link Mutable}) is mutable, the class is designed
+ * to be immutable from the outside. This is because it is difficult to manipulate grouped stages directly without
+ * breaking the invariants of the class, so it is better to be sure it is not modified after it is calculated.
+ */
+public abstract class GroupedStages {
+
+  public static final Comparator<MailboxSendNode> STAGE_COMPARATOR = Comparator.comparing(BasePlanNode::getStageId);
+  public static final Comparator<SortedSet<MailboxSendNode>> GROUP_COMPARATOR
+      = Comparator.comparing(group -> group.first().getStageId());
+
+  public abstract boolean containsStage(MailboxSendNode stage);
+
+  /**
+   * Returns the group of equivalent stages that contains the given stage.
+   *
+   * The set is sorted by the stage id.
+   */
+  public abstract SortedSet<MailboxSendNode> getGroup(MailboxSendNode stage)
+      throws NoSuchElementException;
+
+  /**
+   * Returns the leaders of each group.
+   *
+   * The leader of a group is the stage with the smallest stage id in the group.
+   */
+  public abstract SortedSet<MailboxSendNode> getLeaders();
+
+  /**
+   * Returns the groups.
+   *
+   * Each set contains the stages that are grouped. These sets are disjoint. The union of these sets is the set of all
+   * stages known by this object.
+   *
+   * The result is sorted by the leader of each group and each group is sorted by the stage id.
+   */
+  public abstract SortedSet<SortedSet<MailboxSendNode>> getGroups();
+
+  @Override
+  public String toString() {
+    String content = getGroups().stream()
+        .map(group ->
+            "[" + group.stream()
+                .map(stage -> Integer.toString(stage.getStageId()))
+                .collect(Collectors.joining(", ")) + "]"
+        )
+        .collect(Collectors.joining(", "));
+
+    return "[" + content + "]";
+  }
+
+  /**
+   * A mutable version of {@link GroupedStages}.
+   */
+  public static class Mutable extends GroupedStages {
+    /**
+     * All groups of stages.
+     *
+     * Although these groups are never empty, a group may contain only one stage if it is not grouped with any other
+     * stage.
+     */
+    private final SortedSet<SortedSet<MailboxSendNode>> _groups = new TreeSet<>(GROUP_COMPARATOR);
+
+    /**
+     * Map from stage to the group of stages it belongs to.
+     */
+    private final IdentityHashMap<MailboxSendNode, SortedSet<MailboxSendNode>> _stageToGroup = new IdentityHashMap<>();
```

Comment:
It is enough, but this map let us implement some operations in constant cost instead of supra linear cost (in terms of number of groups). For example, in order to find if a stage is contained in a `GroupedStages` object we would need to iterate over all _groups (linear cost) and for each group ask if the stage is contained or not (O(logn) cost). But thanks to this map the lookup cost is O(1) (and super small, given we don't even need to calculate hashes because we use identity maps)

Reply:
You are right, but we are reducing from `O(SlogN)` to `O(1)` and the complexity added doesn't seem to be that large, right?